### PR TITLE
Provider Agent: postpone instead of outright rejecting negotiations due to concurrency limit

### PR DESCRIPTION
**Pull request by mfranciszkiewicz**
* postponed negotiation timeout: 24h

### CODE REVIEW

File: agent/provider/src/market/provider_market.rs [hunk 1]
```diff
@@ -594,6 +658,55 @@ impl Handler<AgreementFinalized> for ProviderMarket {
...
+                    Some(sub) => sub.clone(),
+                    None => {
+                        log::info!(
+                            "Ignoring postponed negotiation: subscription [{}] no longer exists.",
+                            subscription_id
+                        );
+                        continue;
+                    }
+                };
+
```
Comment (tworec) [inline agent/provider/src/market/provider_market.rs:691]:
```suggestion
                    log::info!(
                        "Taking again postponed negotiation of Proposal [{}] for subscription [{}].",
                        proposal.proposal_id()?,
                        subscription_id
                    );
```

File: agent/provider/src/market/provider_market.rs [hunk 2]
```diff
@@ -230,17 +251,37 @@ impl ProviderMarket {
...
+            log::info!(
+                "Negotiation for Requestor [{}] and subscription [{}] is postponed.",
+                issuer_id,
+                subscription.preset.name,
+            );
+            // update a postponed negotiation with the latest proposal
+            myself
+                .send(PostponeNegotiation {
+                    subscription_id: subscription_id.clone(),
+                    proposal: demand.clone(),
```
[the remark this answers is not in the document]
Comment (mfranciszkiewicz) [reply agent/provider/src/market/provider_market.rs:277]: `demand` is a reference so `.clone()` is required

File: agent/provider/src/market/provider_market.rs [hunk 3]
```diff
@@ -594,6 +658,55 @@ impl Handler<AgreementFinalized> for ProviderMarket {
                 error,
             );
         }
+
+        let myself = ctx.address();
+        let market_api = self.market_api.clone();
+        let subscriptions = self.offer_subscriptions.clone();
+        let negotiations = std::mem::replace(&mut self.postponed_negotiations, HashMap::new());
```
[the remark this answers is not in the document]
Comment (mfranciszkiewicz) [reply agent/provider/src/market/provider_market.rs:665]: Not really, this PR was a draft

File: agent/provider/src/market/provider_market.rs (hunk 1 above)
Comment (mfranciszkiewicz) [reply agent/provider/src/market/provider_market.rs:691]: `proposal.proposal_id()?` - `?` isn't allowed here

File: agent/provider/src/market/provider_market.rs [hunk 4]
```diff
@@ -594,6 +658,55 @@ impl Handler<AgreementFinalized> for ProviderMarket {
...
+                {
+                    log::error!(
+                        "Error processing postponed negotiation for subscription [{}]: {:?}",
+                        subscription_id,
+                        error
+                    )
+                }
+            }
+        };
+        ctx.spawn(fut.into_actor(self));
```
[the remark this answers is not in the document]
Comment (mfranciszkiewicz) [reply agent/provider/src/market/provider_market.rs:708]: Could you provide an example?

File: agent/provider/src/market/provider_market.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (tworec) [reply agent/provider/src/market/provider_market.rs:277]: ok,sorry

File: agent/provider/src/market/provider_market.rs (hunk 1 above)
Comment (tworec) [reply agent/provider/src/market/provider_market.rs:699]: your change is even better than my proposition

File: agent/provider/src/market/provider_market.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (tworec) [reply agent/provider/src/market/provider_market.rs:716]: see `ReSubscribe` handler by @prekucki

[the remark this answers is not in the document]
Comment (mfranciszkiewicz) [reply agent/provider/src/market/provider_market.rs:716]: Sorry but I don't see how it fits this particular case - values are still cloned and moved

**Comment by tworec** [comment]
Together with @mfranciszkiewicz we've decided to solve without Proposal postponing. Provider will unsubscribe and subscribe again all active Offers upon `AgreementFinalized`

Closed via #475

**Pull request closed without merging**